### PR TITLE
Fix Path/str-discrepancy in FontManager.addpath and improve documentation

### LINKED ISSUE #22582: [Bug]: FontManager.addfont doesn't accept pathlib.Path of TTF font
### Bug summary

The [`FontManager.addfont()`](https://github.com/matplotlib/matplotlib/blob/710fce3df95e22701bd68bf6af2c8adbc9d67a79/lib/matplotlib/font_manager.py#L1077-L1094) method is documented as accepting a path-like, but if a path-like for a .ttf font file is passed it raises `TypeError`

### Code for reproduction

```python
from pathlib import Path
from matplotlib import get_data_path, __version__
from matplotlib.font_manager import fontManager

path = Path(get_data_path()) / "fonts" / "ttf" / "cmr10.ttf"

print("matplotlib", __version__)
fontManager.addfont(str(path))  # Works
print("str ok")
fontManager.addfont(path)  # TypeError
print("path ok")
```

### Actual outcome

```
matplotlib 3.5.1
str ok
Traceback (most recent call last):
  File "/home/twm/dev/roofs/repro.py", line 10, in <module>
    fontManager.addfont(path)  # TypeError
  File "/home/twm/dev/roofs/venv/lib/python3.9/site-packages/matplotlib/font_manager.py", line 1092, in addfont
    font = ft2font.FT2Font(path)
TypeError: First argument must be a path or binary-mode file object
```

### Expected outcome

```
matplotlib 3.5.1
str ok
path ok
```

### Additional information

It looks like the simplest fix would be to convert path-likes to `str` since [`FT2Font` is implemented in C++](https://github.com/matplotlib/matplotlib/blob/9197433d50ed77c37a2d5fcb4e0a3f4b26112c8e/src/ft2font.cpp).

### Operating system

Ubuntu 20.04

### Matplotlib Version

3.5.1

### Matplotlib Backend

TkAgg

### Python version

3.9.5

### Jupyter version

N/A

### Installation

pip

### LINKED ISSUE #22586: [Doc]: FontManager docs should mention the singleton is named `fontManager`
### Documentation Link

https://matplotlib.org/devdocs/api/font_manager_api.html

### Problem

The reference documentation for the `FontManager` class says that there is a singleton:

> This module provides a single [FontManager](https://matplotlib.org/3.5.1/api/font_manager_api.html#matplotlib.font_manager.FontManager) instance that can be shared across backends and platforms.

However, it doesn't mention that this singleton is named `fontManager`.

### Suggested improvement

* The docstring should clarify that the singleton is named `fontManager`, rather than implying that the `FontManager` class is a singleton:

    ```restructuredtext
    This module provides a single `FontManager` instance, `fontManager`, that is shared across backends and platforms.
    ```

* The `fontManager` instance should be documented as a global variable:

    ```restructuredtext
    .. data:: fontManager

        The singleton instance of :class:`FontManager`
    ```

### PR DESCRIPTION
**Pull request by oscargus**
## PR Summary

Closes #22582
Closes #22586

Always convert input to `str` first.

Also update the documentation a bit.

Cannot really see how this can be tested as I do not know of any ttf-font locations on the CI... Edit: I found a ttf-font in the test data, so at least it should be possible to test passing a Path without errors. Edit 2: Test added.

Also added a validation function for `font.stretch`, so now rcParam-values can also be ints.

Kept it as four separate commits, but can of course merge to a single.

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/font_manager.py [hunk 1]
```diff
@@ -838,7 +851,8 @@ def set_weight(self, weight):
         Set the font weight.  May be either a numeric value in the
         range 0-1000 or one of 'ultralight', 'light', 'normal',
         'regular', 'book', 'medium', 'roman', 'semibold', 'demibold',
-        'demi', 'bold', 'heavy', 'extra bold', 'black'
+        'demi', 'bold', 'heavy', 'extra bold', 'black'.
+        Default: :rc:`font.weight`
```
Comment (QuLogic) [inline lib/matplotlib/font_manager.py:855]: Should probably change this to numpydoc.

File: lib/matplotlib/font_manager.py [hunk 2]
```diff
@@ -857,6 +871,7 @@ def set_stretch(self, stretch):
         'extra-condensed', 'condensed', 'semi-condensed', 'normal',
         'semi-expanded', 'expanded', 'extra-expanded' or
         'ultra-expanded', or a numeric value in the range 0-1000.
+        Default: :rc:`font.stretch`
```
Comment (QuLogic) [inline lib/matplotlib/font_manager.py:874]: Ditto.

File: lib/matplotlib/font_manager.py (hunk 1 above)
Comment (oscargus) [reply lib/matplotlib/font_manager.py:855]: Yes, I was a bit lazy as I couldn't really figure out what the right way to transform "numeric value in the range 0-1000" in a sensible way to that format. float and then mention the range below?

Comment (QuLogic) [reply lib/matplotlib/font_manager.py:855]: That's not a type; just put it in the text description.

File: lib/matplotlib/tests/test_font_manager.py [hunk 3]
```diff
@@ -176,6 +176,13 @@ def test_user_fonts_linux(tmpdir, monkeypatch):
     _get_fontconfig_fonts.cache_clear()
 
 
+def test_addfont():
+    font_test_file = 'mpltest.ttf'
+    path = Path(__file__).parent / font_test_file
+    # Add font using Path, which should not produce an error. See #22582
+    fontManager.addfont(path)
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_font_manager.py:183]: Wording slightly more to the point:
```suggestion
def test_addfont_as_path():
    """Smoke test that addfont() accepts pathlib.Path."""
    font_test_file = 'mpltest.ttf'
    path = Path(__file__).parent / font_test_file
    fontManager.addfont(path)
```

**Review by timhoffm — COMMENTED** [review]
This is good, only some stylistic comments.

File: lib/matplotlib/tests/test_font_manager.py (hunk 3 above)
Comment (oscargus) [reply lib/matplotlib/tests/test_font_manager.py]: Thanks! Fixed that.

File: lib/matplotlib/font_manager.py [hunk 4]
```diff
@@ -627,32 +631,33 @@ class FontProperties:
...
     - family: A list of font names in decreasing order of priority.
       The items may include a generic font family name, either
-      'sans-serif' (default), 'serif', 'cursive', 'fantasy', or 'monospace'.
+      'sans-serif', 'serif', 'cursive', 'fantasy', or 'monospace'.
       In that case, the actual font to be used will be looked up
-      from the associated rcParam.
+      from the associated rcParam. Default: :rc:`font.family`
 
-    - style: Either 'normal' (default), 'italic' or 'oblique'.
+    - style: Either 'normal', 'italic' or 'oblique'.
```
[the remark this answers is not in the document]
Comment (oscargus) [reply lib/matplotlib/font_manager.py:634]: That would make sense, but I think it is better to try to get the information from the setmethods at a later stage.

File: lib/matplotlib/font_manager.py [hunk 5]
```diff
@@ -11,6 +11,10 @@
 The design is based on the `W3C Cascading Style Sheet, Level 1 (CSS1)
 font specification <http://www.w3.org/TR/1998/REC-CSS2-19980512/>`_.
 Future versions may implement the Level 2 or 2.1 specifications.
+
+.. data:: fontManager
+
+    The singleton instance of `FontManager`.
```
[the remark this answers is not in the document]
Comment (oscargus) [reply lib/matplotlib/font_manager.py]: Thanks! Fixed it.

**Comment by oscargus** [comment]
I changed based on the comments.

However, there is still the same issue as in https://github.com/matplotlib/matplotlib/pull/22556#discussion_r821544848

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/font_manager.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
 """
 A module for finding, managing, and using fonts across platforms.
 
-This module provides a single `FontManager` instance that can
+This module provides a single `FontManager` instance, ``fontManager``, that can
```
Comment (oscargus) [inline lib/matplotlib/font_manager.py:4]: I did not manage to link to `fontManager` when the definition is in `font_manager_api.rst`.

File: lib/matplotlib/font_manager.py [hunk 7]
```diff
@@ -1091,6 +1113,9 @@ def addfont(self, path):
         ----------
         path : str or path-like
         """
+        # Convert to string in case of a path as
+        # afmFontProperty and FT2Font expect this
+        path = str(path)
```
Comment (anntzer) [inline lib/matplotlib/font_manager.py:1118]: I think this should use os.fsdecode instead? (to handle "arbitrary" pathlikes)

Comment (oscargus) [reply lib/matplotlib/font_manager.py:1118]: Seems to be a better way indeed! Thanks!

**Review by QuLogic — APPROVED** [review]

**Comment by twm** [comment]
Thank you all very much for the quick resolution!

**Pull request merged**